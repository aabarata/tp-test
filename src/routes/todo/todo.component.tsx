import { useContext, useEffect, useMemo, useState } from "react";
import { Link, useParams } from "react-router-dom";
import { TodoContext } from "../../context/todo.context";
import { UserContext } from "../../context/user.context";
import { Todo } from "../../@types/todo.d";
import { User } from "../../@types/user.d";
import { getPriorityLabel } from "../../business/todo.helpers";
import Button from "@mui/material/Button";
import ArrowBackIcon from "@mui/icons-material/ArrowBack";

const TodoRoute = () => {
  const { id } = useParams();
  const { getTodoById } = useContext(TodoContext);
  const { getUserByUUID } = useContext(UserContext);
  const [todo, setTodo] = useState<Todo | undefined>(undefined);
  const [assignedUser, setAssignedUser] = useState<User | undefined>(undefined);

  useEffect(() => {
    if (id) {
      const todoById = getTodoById(id);
      setTodo(todoById);
      if (
        todoById &&
        "assignedUserUUID" in todoById &&
        typeof todoById.assignedUserUUID === "string"
      ) {
        const user = getUserByUUID(todoById.assignedUserUUID);
        setAssignedUser(user);
      }
    }
  }, [id, getTodoById, getUserByUUID]);

  const getCreatedAtLabel = (createdAt: number) => {
    return new Date(createdAt).toLocaleDateString("en-US");
  };

  const getUserName = useMemo(() => {
    return assignedUser
      ? `${assignedUser.name.title} ${assignedUser.name.first} ${assignedUser.name.last}`
      : "Unassigned";
  }, [assignedUser]);

  return (
    <div className="container mx-auto pt-6">
      {!todo ? (
        <div>Sorry, was not possible to find the todo with the id {id}</div>
      ) : (
        <div className="w-2/4">
          <div>
            <h1 className="text-xl font-bold">{todo.name}</h1>
            <span className="text-xs text-gray-400 font-semibold">
              {todo.id}
            </span>
          </div>
          <div className="mt-4">
            <label className="text-lg font-semibold mb-2">Notes:</label>
            <p className="bg-gray-200 py-1 px-3 italic">{todo.notes}</p>
          </div>
          <div className="mt-4">
            <label className="text-lg font-semibold mb-2">Priority:</label>
            <p className="bg-gray-200 py-1 px-3 italic">
              {getPriorityLabel(todo.priority)}
            </p>
          </div>
          <div className="mt-4">
            <label className="text-lg font-semibold mb-2">Assigned to:</label>
            <div className="flex flex-row items-center bg-gray-200 py-1 px-3 italic">
              <img
                className="mr-2 rounded-full"
                alt="assigned user"
                width={50}
                height={50}
                src={assignedUser?.picture.thumbnail}
              />
              <p>{getUserName}</p>
            </div>
          </div>
          <div className="mt-4">
            <label className="text-lg font-semibold mb-2">Status:</label>
            <p className="bg-gray-200 py-1 px-3 italic">
              {todo.completed ? "Completed" : "Incomplete"}
            </p>
          </div>
          <div className="mt-4">
            <label className="text-lg font-semibold mb-2">Created at:</label>
            <p className="bg-gray-200 py-1 px-3 italic">
              {getCreatedAtLabel(todo.createdAt)}
            </p>
          </div>
          <Link to="/">
            <Button
              startIcon={<ArrowBackIcon />}
              sx={{ color: "white", marginTop: "24px" }}
              variant="contained"
            >
              Back
            </Button>
          </Link>
        </div>
      )}
    </div>
  );
};

export default TodoRoute;
