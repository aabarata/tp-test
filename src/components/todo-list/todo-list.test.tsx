import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Todo } from "../../@types/todo.d";
import TodoList from "./todo-list.component";

describe("todoCard", () => {
  const mockedTodos: Todo[] = [
    {
      assignedUserUUID: "33b250e0-fbaa-4a24-bd80-02f41cbe6c65",
      completed: false,
      createdAt: 1699208593120,
      id: "9df5d6db-3cc2-4c86-919b-b781f0020c73",
      name: "todo 1",
      notes: "Random notes 1",
      priority: 2,
    },
    {
      name: "todo 2",
      notes: "Random notes 2",
      priority: 1,
      assignedUserUUID: "9310d272-78b3-47c3-a989-0f5b0f7bb302",
      completed: false,
      createdAt: 1699208605997,
      id: "3b6d64ca-331b-41d9-9bd3-40730b798b87",
    },
  ];
  it("Should match the snapshot", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/"]}>
        <TodoList todos={mockedTodos} />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
  it("Should display all the todos", () => {
    render(
      <MemoryRouter initialEntries={["/"]}>
        <TodoList todos={mockedTodos} />
      </MemoryRouter>
    );
    expect(screen.getByText("todo 1")).toBeInTheDocument();
    expect(screen.getByText("todo 2")).toBeInTheDocument();
  });
});
