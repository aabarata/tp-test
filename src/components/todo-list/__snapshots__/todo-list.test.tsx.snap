// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todoCard Should match the snapshot 1`] = `
<div>
  <div
    class="rounded-lg p-3 list bg-blue-col"
  >
    <div
      class="flex flex-row bg-white px-1.5 py-2 rounded mb-2"
    >
      <div
        class="flex flex-row items-center justify-center mr-2"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          data-testid="card-action-status"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <div
        class="flex flex-col flex-1"
      >
        <div
          class="flex flex-row items-center justify-between mb-1"
        >
          <span
            class="flex flex-row items-center font-bold true"
          >
            <span
              class="mr-2"
              data-testid="card-name"
            >
              todo 1
            </span>
            <a
              href="/todos/9df5d6db-3cc2-4c86-919b-b781f0020c73"
            >
              <button
                aria-label="open todo"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-yo6ymh-MuiButtonBase-root-MuiIconButton-root"
                data-testid="card-action-details"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                  data-testid="VisibilityIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </a>
            <button
              aria-label="edit todo"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-yo6ymh-MuiButtonBase-root-MuiIconButton-root"
              data-testid="card-action-update"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="delete todo"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-yo6ymh-MuiButtonBase-root-MuiIconButton-root"
              data-testid="card-action-remove"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </span>
          <span
            class="text-xs font-semibold"
            data-testid="card-priority"
            style="color: orange;"
          >
            MEDIUM
          </span>
        </div>
        <div
          class="flex flex-row items-center justify-between"
        >
          <span
            class="text-sm text-gray-400"
            data-testid="card-created-at"
          >
            Created at: 
            11/5/2023
          </span>
          <img
            alt="assigned user"
            class="rounded-full"
            data-testid="card-assigned-photo"
            height="20"
            width="20"
          />
        </div>
      </div>
      <div>
        <div />
      </div>
    </div>
    <div
      class="flex flex-row bg-white px-1.5 py-2 rounded "
    >
      <div
        class="flex flex-row items-center justify-center mr-2"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          data-testid="card-action-status"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
      <div
        class="flex flex-col flex-1"
      >
        <div
          class="flex flex-row items-center justify-between mb-1"
        >
          <span
            class="flex flex-row items-center font-bold true"
          >
            <span
              class="mr-2"
              data-testid="card-name"
            >
              todo 2
            </span>
            <a
              href="/todos/3b6d64ca-331b-41d9-9bd3-40730b798b87"
            >
              <button
                aria-label="open todo"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-yo6ymh-MuiButtonBase-root-MuiIconButton-root"
                data-testid="card-action-details"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                  data-testid="VisibilityIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </a>
            <button
              aria-label="edit todo"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-yo6ymh-MuiButtonBase-root-MuiIconButton-root"
              data-testid="card-action-update"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="delete todo"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-yo6ymh-MuiButtonBase-root-MuiIconButton-root"
              data-testid="card-action-remove"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </span>
          <span
            class="text-xs font-semibold"
            data-testid="card-priority"
            style="color: green;"
          >
            LOW
          </span>
        </div>
        <div
          class="flex flex-row items-center justify-between"
        >
          <span
            class="text-sm text-gray-400"
            data-testid="card-created-at"
          >
            Created at: 
            11/5/2023
          </span>
          <img
            alt="assigned user"
            class="rounded-full"
            data-testid="card-assigned-photo"
            height="20"
            width="20"
          />
        </div>
      </div>
      <div>
        <div />
      </div>
    </div>
  </div>
</div>
`;
